### DISCUSSION TITLE
Exit code 1

### DISCUSSION THREAD

**Discussion by christozlowski** [IBM/mac-ibm-notifications#217] [Q&A]
We have a Jamf policy for most of our Macs that runs a script and launches a notification with IBM Notifier. On one device in particular, IBM Notifier does not launch the notification and returns an exit code of 1, which is an internal error according to the wiki.

Is an internal error a bug or is there something we can do to troubleshoot this?

Here's an example of how we're using it:

```
	prompt_user() {
		button=$( "${ibm_notifier_path}/Contents/MacOS/${ibm_notifier_binary}" \
		-type "popup" \
		-bar_title "Reminder" \
		-subtitle "$main_text" \
		-icon_path "$icon" \
		-main_button_label "No" \
		-secondary_button_label "Yes" \
		"${help_info[@]}" \
		-timeout 120 \
		"${sound[@]}" \
		-position center \
		-always_on_top )
		
		echo "$?"
	}
```

This problem is on macOS 14.5 with both version 3.1.0 and 3.2.0 of IBM Notifier.

**Comment by christozlowski** [comment]
Here is another example in a different script we use that also has an internal error. 

```
"${NA_PATH}" \
	-type "$type" \
	-title "$title" \
	-subtitle "$subtitle" \
	-main_button_label "$btn_label" \
	-main_button_cta_type "$btn_type" \
	-main_button_cta_payload "$btn_payload" &
```

**Comment by SMartorelli** [comment]
Hi @christozlowski 👋 it's hard to debug an issue in a script without the complete script. There might be something in one of the used variables that cause this behaviour. Could you please share the complete script that cause the error removing any private/org information so that I can take a look at it? you can also directly send it to [me](https://macadmins.slack.com/team/U01UB6AS0HJ) on Slack in the [Mac Admins workspace](https://www.macadmins.org). Thanks!

**Reply by christozlowski** [reply]
Thank you for looking into this! Here is an open source version of what we're using: https://github.com/sgmills/PrivilegesDemoter
It occurs in [PrivilegesDemoter.sh](https://github.com/sgmills/PrivilegesDemoter/blob/f7fc9a133ff2c7c5a46b2da72e135ed47c76c5e4/pkg/PrivilegesDemoter.sh)

It does appear to be happening in any script though, not just that one.

Here is another script where it occurs. This one is run from a Jamf policy where we can change the parameters.
https://gist.github.com/christozlowski/4c6d2f6e9f98bc82424c83cd3e15ab2f

**Reply by SMartorelli** [reply]
I'm not able to reproduce the issue with the linked scripts. Both seems to work just fine during my tries. 
Are you sure you're getting the exit code 1 from IBM Notifier and not from other section of the script? For example a different installation location for IBM Notifier might also cause it. 
I don't see any report here or in PrivilegesDemoter repo about unexpected exit code 1 from the notifier.

**Reply by christozlowski** [reply]
I'm fairly certain it's from IBM Notifier. The logs have "Error: Unexpected exit code 1 returned from prompt", followed by the user and machine UDID which comes from line 506 in PrivilegesDemoter. Same with line 107 in the notification script.

**Reply by christozlowski** [reply]
I might have another piece of the puzzle. I ran this on the affected device: `/Applications/IBM\ Notifier.app/Contents/MacOS/IBM\ Notifier --v -type "banner" -title "Test Title" -subtitle "test subtitle" -main_button_label "OK" -main_button_cta_type "none"
`
The output was: 

> Notification object successfully created from arguments list.
> Rich notification authorization request ended with error: Notifications are not allowed for this application

We do have a configuration profile to allow notifications for IBM Notifier.

**Reply by SMartorelli** [reply]
Got it, this is a known issue in macOS with the deployment of notification payload profiles. Please check out [our solution](https://github.com/IBM/mac-ibm-notifications/wiki/Known-Issues-And-Fixes#4-notification-agent-rich-notification-authorization-request-ended-with-error-notifications-are-not-allowed-for-this-application) to this error in the wiki.

**Reply by christozlowski** [reply]
Okay, that appears to have worked for alerts and banners! However, popups still do not work. 

`/Applications/IBM\ Notifier.app/Contents/MacOS/IBM\ Notifier --v -type popup -bar_title "Test" -subtitle "test text" -main_button_label "No" -secondary_button_label "Yes" -position center`

> Notification object successfully created from arguments list.
Apple script returned error: {
  NSAppleScriptErrorAppName = "IBM Notifier Popup";
  NSAppleScriptErrorBriefMessage = "2024-08-01 08:51:09.118 defaults[38937:594The domain/default pair of (kCFPreferencesAnyApplication, com.apple.sound.uiaudio.enabled) does not exist";
  NSAppleScriptErrorMessage = "2024-08-01 08:51:09.118 defaults[38937:5944411]The domain/default pair of (kCFPreferencesAnyApplication, com.apple.sound.uiaudio.enabled) does not exist";
  NSAppleScriptErrorNumber = 1;
  NSAppleScriptErrorRange = "NSRange: {0, 66}";
}

**Reply by SMartorelli** [reply]
That it's just some verbosity coming from Sonoma. Just avoid to use --v (verbose mode) if you're not troubleshooting errors.

**Reply by christozlowski** [reply]
I thought verbose mode might help troubleshoot the problem. No popup appears and that is the output in verbose mode. Still with exit code 1

**Reply by SMartorelli** [reply]
I don't see any issue with the invocation you posted
`/Applications/IBM\ Notifier.app/Contents/MacOS/IBM\ Notifier -type popup -bar_title "Test" -subtitle "test text" -main_button_label "No" -secondary_button_label "Yes" -position center`
Just tried it on my device with the IBM Notifier version from last release 3.2.0 (126).
Are you trying it directly from your device or through Jamf script? Not sure why the popup doesn't appear in your case. Do you hear the popup sound?

**Comment by christozlowski** [comment]
I'm going to close this. We ended up giving up and wiped the problem device and it went away. Thank you helping!
